connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/util/RemoteSparkSession.scala

```diff
@@ -0,0 +1,198 @@
+ */
+package org.apache.spark.sql.connect.client.util
+
+import java.io.{BufferedOutputStream, File}
+
+import scala.io.Source
+
+import org.scalatest.BeforeAndAfterAll
+import sys.process._
+
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.connect.client.SparkConnectClient
+import org.apache.spark.sql.connect.common.config.ConnectCommon
+
+/**
+ * An util class to start a local spark connect server in a different process for local E2E tests.
+ * It is designed to start the server once but shared by all tests. It is equivalent to use the
+ * following command to start the connect server via command line:
+ *
+ * {{{
+ * bin/spark-shell \
+ * --jars `ls connector/connect/server/target/**/spark-connect*SNAPSHOT.jar | paste -sd ',' -` \
+ * --conf spark.plugins=org.apache.spark.sql.connect.SparkConnectPlugin
+ * }}}
+ *
+ * Set system property `SPARK_HOME` if the test is not executed from the Spark project top folder.
+ * Set system property `DEBUG_SC_JVM_CLIENT=true` to print the server process output in the
+ * console to debug server start stop problems.
+ */
+object SparkConnectServerUtils {
+  // System properties used for testing and debugging
+  private val SPARK_HOME = "SPARK_HOME"
+  private val ENV_DEBUG_SC_JVM_CLIENT = "DEBUG_SC_JVM_CLIENT"
+
+  private val sparkHome = System.getProperty(SPARK_HOME, fileSparkHome())
+  private val isDebug = System.getProperty(ENV_DEBUG_SC_JVM_CLIENT, "false").toBoolean
+
+  // Log server start stop debug info into console
+  // scalastyle:off println
+  private[connect] def debug(msg: String): Unit = if (isDebug) println(msg)
+  // scalastyle:on println
+  private[connect] def debug(error: Throwable): Unit = if (isDebug) error.printStackTrace()
+
+  // Server port
+  private[connect] val port = ConnectCommon.CONNECT_GRPC_BINDING_PORT + util.Random.nextInt(1000)
+
+  @volatile private var stopped = false
+
+  private lazy val sparkConnect: Process = {
+    debug("Starting the Spark Connect Server...")
+    val jar = findSparkConnectJar
+    val builder = Process(
+      new File(sparkHome, "bin/spark-shell").getCanonicalPath,
+      Seq(
+        "--jars",
+        jar,
+        "--driver-class-path",
+        jar,
+        "--conf",
+        "spark.plugins=org.apache.spark.sql.connect.SparkConnectPlugin",
+        "--conf",
+        s"spark.connect.grpc.binding.port=$port"))
+
+    val io = new ProcessIO(
+      // Hold the input channel to the spark console to keep the console open
+      in => new BufferedOutputStream(in),
+      // Only redirect output if debug to avoid channel interruption error on process termination.
+      out => if (isDebug) Source.fromInputStream(out).getLines.foreach(debug),
+      err => if (isDebug) Source.fromInputStream(err).getLines.foreach(debug))
+    val process = builder.run(io)
+
+    // Adding JVM shutdown hook
+    sys.addShutdownHook(kill())
+    process
+  }
+
+  def start(): Unit = {
+    assert(!stopped)
+    sparkConnect
```

Comment:
Can we do this w/ a proper `start` and `stop`? For example, if you call `kill`, then it will start a process, and then kill.

Reply:
I tried to stop the server properly by sending `:quit` to the shell process, but it works 20% of the runs. As it is too unreliable I has to use the kill. Do you know some better way to stop the shell properly? Thanks a lot.